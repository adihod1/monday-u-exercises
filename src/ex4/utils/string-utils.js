export function capitalizeText(text) {
    return text
        .split(' ')
        .map((s) => s[0].toUpperCase() + s.slice(1))
        .join(' ');
}


export function generateUniqueID() { 
    return Math.random().toString(36).substr(2, 9);
}