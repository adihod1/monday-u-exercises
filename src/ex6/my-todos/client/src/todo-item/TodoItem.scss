.todo-checkbox {
  top: 12px;
  position: relative;
}

.todo-message {
  position: relative;
  display: inline-block;
  background: transparent;
  border: 0;
  width: 100%;
}

.todo-message.editable {
  font-size: 16px;
  font-weight: bolder;
  background: white;
}
