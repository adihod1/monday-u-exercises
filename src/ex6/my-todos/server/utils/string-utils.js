
function capitalizeText(text) {
    return text
        .split(' ')
        .map((s) => s[0].toUpperCase() + s.slice(1))
        .join(' ');
}

function generateUniqueID() {
    return Math.random().toString(36).substr(2, 9);
}

module.exports = {
    capitalizeText,
    generateUniqueID
}