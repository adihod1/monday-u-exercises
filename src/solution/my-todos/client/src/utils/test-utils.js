export const todosList = [
    {
        id: 1,
        name: 'do homework',
        type: 'text',
        message: 'do homework',
        status: 0
    },
    {
        id: 2,
        name: 'clean the room',
        type: 'text',
        message: 'clean the room',
        status: 1
    },
    {
        id: 3,
        name: 'call a friend',
        type: 'text',
        message: 'call a friend',
        status: 0
    },
    {
        id: 4,
        name: "hjkh",
        type: "text",
        message: "hjkh",
        status: 0
    },
    {
        id: 5,
        name: "bake a cake by the day after tomorrow",
        type: "text",
        message: "bake a cake by the day after tomorrow",
        status: 0
    },
    {
        id: 6,
        name: "84",
        type: "pokemon",
        message: "Catch #84 Doduo the Normal/Flying type pokemon",
        pokemon: {
            pokemon_id: "84",
            pokemon_name: "doduo",
            pokemon_image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/84.png",
            pokemon_type: "[\"Normal\",\"Flying\"]"
        },
        status: 1
    },
    {
        id: 7,
        name: "4",
        type: "pokemon",
        message: "Catch #4 Charmander the Fire type pokemon",
        pokemon: {
            pokemon_id: "4",
            pokemon_name: "charmander",
            pokemon_image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png",
            pokemon_type: "[\"Fire\"]"
        },
        status: 0
    },
    {
        id: 8,
        name: "25",
        type: "pokemon",
        message: "Catch #25 Pikachu the Electric type pokemon",
        pokemon: {
            pokemon_id: "25",
            pokemon_name: "pikachu",
            pokemon_image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/25.png",
            pokemon_type: "[\"Electric\"]"
        },
        status: 0
    },
    {
        id: 9,
        name: "1",
        type: "pokemon",
        message: "Catch #1 Bulbasaur the Grass/Poison type pokemon",
        pokemon: {
            pokemon_id: "1",
            pokemon_name: "bulbasaur",
            pokemon_image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
            pokemon_type: "[\"Grass\",\"Poison\"]"
        },
        status: 0
    },
    {
        id: 10,
        name: "7",
        type: "pokemon",
        message: "Catch #7 Squirtle the Water type pokemon",
        pokemon: {
            pokemon_id: "7",
            pokemon_name: "squirtle",
            pokemon_image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/7.png",
            pokemon_type: "[\"Water\"]"
        },
        status: 0
    }
];


export const emptyStringInitialState = {
    value: ''
};

export const emptyArrayInitialState = {
    todos: []
};