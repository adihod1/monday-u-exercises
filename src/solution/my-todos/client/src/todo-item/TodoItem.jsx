import React, { useState, useCallback, useRef, useMemo, useLayoutEffect } from "react";
import { faPen, faTrash, faCheck } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import "./TodoItem.scss";
import ItemClient from "../api/itemClient";
import { useClickOutside } from '../hooks/useClickOutside'

const itemClient = new ItemClient();

const messageProps = (editable, onApproveClick, onTodoClick) => {
    const props = {
        disabled: !editable,
        className: `todo-message ${editable ? 'editable' : ''}`,
    }
    if (editable) {
        props.onKeyUp = e => e.key === 'Enter' ? onApproveClick() : null
    } else {
        props.onClick = onTodoClick
    }
    return props
}

const saveCheckbox = async (id, todo, setChecked) => {
    todo.checked = !todo.checked
    await editTodo(id, todo)
    setChecked(todo.checked)
}

const editItem = async (id, todo, item) => {
    todo.item = item
    const newItem = await editTodo(id, todo)
    return newItem.body
}

const editTodo = async (id, todo) => {
    return await itemClient.editTodo(id, {
        type: todo.type,
        item: todo.item,
        checked: todo.checked,
        done_timestamp: Date.now()
    })
}

const TodoItem = ({ todo, setSpinner, changePendingTodos, refreshState }) => {
    const [checked, setChecked] = useState(!!todo.checked)
    const [item, setItem] = useState(todo.item)
    const [message, setMessage] = useState(todo.message)
    const [editable, setEditable] = useState(false)
    const inputRef = useRef();
    const editRef = useRef();
    useClickOutside(editRef, () => setEditable(false));


    const setTodo = (todo) => {
        setItem(todo.item)
        setChecked(todo.checked)
        setMessage(todo.message)
    }

    const pic = todo.type === 'pokemon' ?
        (<img className="pokemon-pic" src={todo.pokemon.image} alt="" />) :
        (<>&emsp;&emsp;</>);

    const onCheckboxChange = async () => {
        setSpinner(true);
        await saveCheckbox(todo.id, todo, setChecked)
        changePendingTodos(checked ? 1 : -1)
        setSpinner(false);
    }

    const onTodoClick = () => {
        alert(todo.message);
    }

    const onDeleteClick = async () => {
        setSpinner(true);
        await itemClient.deleteTodo(todo.id);
        refreshState({});
        setSpinner(false);
    }

    const onEditClick = async () => {
        setEditable(!editable)
    }

    const onApproveClick = async () => {
        setSpinner(true);
        const newItem = await editItem(todo.id, todo, item)
        setTodo(newItem)
        setEditable(false);
        setSpinner(false);
    }

    useLayoutEffect(() => {
        inputRef.current?.focus()
    }, [editable])

    const editBtn = todo.type === 'text' ?
        (
            <>
                <span
                    onClick={onApproveClick}
                    className={`icon-approve ${editable ? '' : 'hidden'}`}>
                    <FontAwesomeIcon
                        className="hover-zoom"
                        icon={faCheck} />
                </span>
                <span
                    onClick={onEditClick}
                    className="icon-edit">
                    <FontAwesomeIcon
                        className="hover-zoom"
                        icon={faPen} />
                </span>
            </>
        ) :
        (<></>);

    return (
        <li ref={editRef}>
            <input
                onChange={onCheckboxChange}
                type="checkbox"
                name="todoCheckbox"
                className="todo-checkbox"
                checked={checked} />
            {pic}
            <input
                ref={inputRef}
                type="text"
                {...messageProps(editable, onApproveClick, onTodoClick)}
                onChange={e => setItem(e.target.value)}
                value={editable ? item : message} />
            {editBtn}
            <span
                onClick={onDeleteClick}
                className="icon-delete">
                <FontAwesomeIcon
                    className="hover-zoom"
                    icon={faTrash} />
            </span>
        </li>
    );
};
export default TodoItem;