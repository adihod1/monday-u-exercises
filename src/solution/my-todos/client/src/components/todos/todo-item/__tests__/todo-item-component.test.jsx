import renderer from "react-test-renderer";
import TodoItem from "../todo-item-component";
import {todosList} from "../../../../utils/test-utils";

test('renders correctly TodoItem ', () => {
    const tree = renderer.create(
        <TodoItem todo={todosList[0]}/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

