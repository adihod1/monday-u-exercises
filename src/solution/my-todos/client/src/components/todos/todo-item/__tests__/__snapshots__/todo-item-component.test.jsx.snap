// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly TodoItem  1`] = `
<li
  className="todo-list-item"
>
  <input
    className="todo-checkbox"
    name="todoCheckbox"
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="todo-item-desc"
    onClick={[Function]}
  >
      
    <input
      className="todo-message "
      disabled={true}
      onChange={[Function]}
      type="text"
      value="do homework"
    />
  </span>
  <span
    className="icon-approve hidden"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-check hover-zoom"
      data-icon="check"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M438.6 105.4C451.1 117.9 451.1 138.1 438.6 150.6L182.6 406.6C170.1 419.1 149.9 419.1 137.4 406.6L9.372 278.6C-3.124 266.1-3.124 245.9 9.372 233.4C21.87 220.9 42.13 220.9 54.63 233.4L159.1 338.7L393.4 105.4C405.9 92.88 426.1 92.88 438.6 105.4H438.6z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </span>
  <span
    className="icon-edit"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-pen hover-zoom"
      data-icon="pen"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M362.7 19.32C387.7-5.678 428.3-5.678 453.3 19.32L492.7 58.75C517.7 83.74 517.7 124.3 492.7 149.3L444.3 197.7L314.3 67.72L362.7 19.32zM421.7 220.3L188.5 453.4C178.1 463.8 165.2 471.5 151.1 475.6L30.77 511C22.35 513.5 13.24 511.2 7.03 504.1C.8198 498.8-1.502 489.7 .976 481.2L36.37 360.9C40.53 346.8 48.16 333.9 58.57 323.5L291.7 90.34L421.7 220.3z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </span>
  <span
    className="icon-delete"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-trash hover-zoom"
      data-icon="trash"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M135.2 17.69C140.6 6.848 151.7 0 163.8 0H284.2C296.3 0 307.4 6.848 312.8 17.69L320 32H416C433.7 32 448 46.33 448 64C448 81.67 433.7 96 416 96H32C14.33 96 0 81.67 0 64C0 46.33 14.33 32 32 32H128L135.2 17.69zM394.8 466.1C393.2 492.3 372.3 512 346.9 512H101.1C75.75 512 54.77 492.3 53.19 466.1L31.1 128H416L394.8 466.1z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </span>
</li>
`;

exports[`renders correctly above limit 1`] = `
<li
  className="todo-list-item"
>
  <input
    className="todo-checkbox"
    name="todoCheckbox"
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="todo-item-desc"
    onClick={[Function]}
  >
      
    <input
      className="todo-message "
      disabled={true}
      onChange={[Function]}
      type="text"
      value="do homework"
    />
  </span>
  <span
    className="icon-approve hidden"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-check hover-zoom"
      data-icon="check"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M438.6 105.4C451.1 117.9 451.1 138.1 438.6 150.6L182.6 406.6C170.1 419.1 149.9 419.1 137.4 406.6L9.372 278.6C-3.124 266.1-3.124 245.9 9.372 233.4C21.87 220.9 42.13 220.9 54.63 233.4L159.1 338.7L393.4 105.4C405.9 92.88 426.1 92.88 438.6 105.4H438.6z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </span>
  <span
    className="icon-edit"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-pen hover-zoom"
      data-icon="pen"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M362.7 19.32C387.7-5.678 428.3-5.678 453.3 19.32L492.7 58.75C517.7 83.74 517.7 124.3 492.7 149.3L444.3 197.7L314.3 67.72L362.7 19.32zM421.7 220.3L188.5 453.4C178.1 463.8 165.2 471.5 151.1 475.6L30.77 511C22.35 513.5 13.24 511.2 7.03 504.1C.8198 498.8-1.502 489.7 .976 481.2L36.37 360.9C40.53 346.8 48.16 333.9 58.57 323.5L291.7 90.34L421.7 220.3z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </span>
  <span
    className="icon-delete"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-trash hover-zoom"
      data-icon="trash"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M135.2 17.69C140.6 6.848 151.7 0 163.8 0H284.2C296.3 0 307.4 6.848 312.8 17.69L320 32H416C433.7 32 448 46.33 448 64C448 81.67 433.7 96 416 96H32C14.33 96 0 81.67 0 64C0 46.33 14.33 32 32 32H128L135.2 17.69zM394.8 466.1C393.2 492.3 372.3 512 346.9 512H101.1C75.75 512 54.77 492.3 53.19 466.1L31.1 128H416L394.8 466.1z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </span>
</li>
`;
